crates/wac-graph/src/graph.rs

```diff
@@ -0,0 +1,1884 @@
+        node: NodeId,
+        /// The kind of the node.
+        kind: String,
+    },
+    /// The node is not an instantiation.
+    #[error("the specified node is not an instantiation")]
+    NodeIsNotAnInstantiation {
+        /// The node that is not an instantiation.
+        node: NodeId,
+    },
+    /// The instance does not export an item with the given name.
+    #[error("instance does not have an export named `{export}`")]
+    InstanceMissingExport {
+        /// The instance node that is missing the export.
+        node: NodeId,
+        /// The export that was missing.
+        export: String,
+    },
+    /// The provided argument name is invalid.
+    #[error("argument name `{name}` is not an import of package `{package}`")]
+    InvalidArgumentName {
+        /// The instantiation node that does not have the provided argument name.
+        node: NodeId,
+        /// The name of the invalid argument.
+        name: String,
+        /// The name of the package that does not have the argument.
+        package: String,
+    },
+    /// The source type does not match the target argument type.
+    #[error("mismatched instantiation argument `{name}`")]
+    ArgumentTypeMismatch {
+        /// The name of the argument.
+        name: String,
+        /// The source of the error.
+        #[source]
+        source: anyhow::Error,
+    },
+    /// The argument has already satisfied.
+    #[error("argument `{name}` has already satisfied")]
+    ArgumentAlreadySatisfied {
+        /// The target node that already has the satisfied argument.
+        node: NodeId,
+        /// The name of the argument that is already satisfied.
+        name: String,
+    },
+    /// The graph contains a cycle.
+    #[error("the graph contains a cycle and cannot be encoded")]
+    GraphContainsCycle {
+        /// The node where the cycle was detected.
+        node: NodeId,
+    },
+    /// The encoding of the graph failed validation.
+    #[error("the encoding of the graph failed validation")]
+    EncodingValidationFailure {
+        /// The source of the validation error.
+        #[source]
+        source: BinaryReaderError,
+    },
+    /// The node cannot be unmarked from export as it is a type definition.
+    #[error("the node cannot be unmarked from export as it is a type definition")]
+    MustExportDefinition,
+    /// An implicit import on an instantiation conflicts with an explicit import node.
+    #[error("an instantiation node implicitly imports an item named `{name}`, but it conflicts with an explicit import node of the same name")]
+    ImplicitImportConfig {
+        /// The existing import node.
+        import: NodeId,
+        /// The instantiation node with the implicit import
+        instantiation: NodeId,
+        /// The name of the conflicting import.
+        name: String,
+    },
+    /// An error occurred while aggregating a type.
+    #[error("failed to aggregate type definition for import `{import}`")]
```

Comment:
I worry that "type aggregation" is going to be meaningless to almost all users. Is there some way we can phrase this that's a bit more meaningful?

Reply:
Agreed, it should be meaningless; let me see if I can word this in a way that makes sense to users.